### PR TITLE
Ensure EvoDB consistency for quorum commitments by storing the best block hash

### PR DESCRIPTION
**Pull request by UdjinM6**
This approach is similar to the one used for chainstate currently.

Ensures that:
- nodes that upgraded after DIP3 has all the data processed correctly (they will crash and require reindex);
- `evodb/` is removed when `blocks/` is removed (otherwise sync will fail due to duplicate records).

~Based on #2536 because I needed `pindex` for this, will rebase later. See 15eb4c1 for actual changes.~

### CODE REVIEW

**Comment by codablock** [comment]
Shouldn't we do this for CEvoDB in general instead of in CQuorumBlockProcessor? CEvoDB is also used by DIP3 and future code.

**Comment by UdjinM6** [comment]
Hmm.. good point.

File: src/llmq/quorums_blockprocessor.cpp [hunk 1]
```diff
@@ -109,10 +110,22 @@ bool CQuorumBlockProcessor::ProcessBlock(const CBlock& block, const CBlockIndex*
 {
     AssertLockHeld(cs_main);
 
+    // Make sure evodb is consistent.
+    // If we already have best block hash saved, the previous block should match it.
+    // If we don't, it's ok too but only before DIP3 activation.
+    uint256 hashBestBlock = uint256();
+    bool fHasHashBestBlock = evoDb.Read(DB_BEST_BLOCK, hashBestBlock);
+    uint256 hashPrevBlock = fHasHashBestBlock ? pindex->pprev->GetBlockHash() : uint256();
```
Comment (gladcow) [inline src/llmq/quorums_blockprocessor.cpp]: May be it is better to check that `pindex` and `pindex->pprev` are not empty before using them?

**Comment by UdjinM6** [comment]
Rebased and reworked to make it more general - moved everything to `ConnectTip`/`DisconnectTip`.

File: src/validation.cpp [hunk 2]
```diff
@@ -2581,6 +2582,23 @@ bool static ConnectTip(CValidationState& state, const CChainParams& chainparams,
         LogPrint("bench", "  - Connect total: %.2fms [%.2fs]\n", (nTime3 - nTime2) * 0.001, nTimeConnectTotal * 0.000001);
         bool flushed = view.Flush();
         assert(flushed);
+
+        // Make sure evodb is consistent.
+        // If we already have best block hash saved, the previous block should match it.
+        // If we don't, it's ok too but only before DIP3 activation.
+        uint256 hashBestBlock = uint256();
```
Comment (codablock) [inline src/validation.cpp]: Why the extra `= uint256()`?

Comment (UdjinM6) [reply src/validation.cpp]: ¯\\\_(ツ)\_/¯

**Comment by UdjinM6** [comment]
Reworked

**Comment by UdjinM6** [comment]
Updated :)

**Comment by codablock** [comment]
Test failures on Travis:
```
stderr:
dashd: evo/evodb.cpp:22: bool CEvoDB::VerifyBestBlock(const uint256&): Assertion `hashBestBlock == hashBlockIndex' failed.
```

**Comment by UdjinM6** [comment]
Surprisingly, `walletbackup.py` became a test for this :D Fixed it.

File: src/evo/evodb.cpp [hunk 3]
```diff
@@ -11,3 +11,20 @@ CEvoDB::CEvoDB(size_t nCacheSize, bool fMemory, bool fWipe) :
...
 }
+
+bool CEvoDB::VerifyBestBlock(const uint256& hash)
+{
+    // Make sure evodb is consistent.
+    // If we already have best block hash saved, the previous block should match it.
+    uint256 hashBestBlock;
+    bool fHasBestBlock = Read(EVODB_BEST_BLOCK, hashBestBlock);
+    uint256 hashBlockIndex = fHasBestBlock ? hash : uint256();
+    assert(hashBestBlock == hashBlockIndex);
```
Comment (codablock) [inline src/evo/evodb.cpp]: Now we got 2 asserts, one here and one by the caller. Is this intended?

**Pull request merged**